### DISCUSSION TITLE
Best practices for working with Keras when switching between Mac and Linux

### DISCUSSION THREAD

**Discussion by jonbry** [rstudio/keras#1432] [Q&A] [answered]
My main computer is a Mac with an M1 chip, and I was recently able to get my hands on PC that has a NVIDIA graphics card that I can put Linux on. I normally use GitHub and the `here` package to keep my projects up to date across platforms, but it seems a bit trickier with Linux and Keras (I should note that this is my first time using Linux). I had a few questions come up that I need a bit of guidance on:

- The Mac is running R 4.3.2, but I noticed the latest [Ubuntu package for R](https://cran.r-project.org/bin/linux/ubuntu/fullREADME.html) is 4.2.  Based on what I read on CRAN, if I want to have 4.3.2 on a Linux computer, I would need to have Debian. I am indifferent between Ubuntu and Debian but TensorFlow/Keras has only been validated for Ubuntu. What would be the best option? I've always kept the same version of R across devices, will having two different versions cause headaches? 

- My main computer will be the Mac and I plan on building larger models on the Linux computer. Any issues just copying the .keras file back to the Mac afterwards? 

Let me know if you have any best practices, advice, or things to watch out for.

Thank you!

**Comment by t-kalinowski** [comment, marked as answer]
Congrats on getting started with Linux! It really is the best environment for doing deep learning.

Some getting started tips:


## Picking a Linux Distribution.

If you'll be using TensorFlow, then Ubuntu is the only officially supported distribution. Unless there is another motivating factor, I recommend using the latest Ubuntu LTS.

That said, in practice, any of the platforms supported by Nvidia CUDA should work: https://developer.nvidia.com/cuda-12-3-2-download-archive?target_os=Linux&target_arch=x86_64. E.g., if you want to use Fedora or RHEL, everything should work still, though you're more likely to encounter some occasional bugs with the TensorFlow stack as it's off the beaten path a little (and you'll also encounter less documentation for the same reason). 


## Installing R: 

- The default R (and Python) versions that ship with the Ubuntu LTS default repositories can start to feel old quickly. To get the latest version, you'll need to install R from somewhere else. Here are some options, sorted roughly in terms of convenience:
 - Use [`rig`](https://github.com/r-lib/rig). 
 - Use the Rstudio pre-built binaries https://github.com/rstudio/r-builds
 - update `/etc/apt/sources.list` as described here: https://cran.r-project.org/bin/linux/ubuntu/fullREADME.html#installing-r
 - Install R from source: https://docs.posit.co/resources/install-r-source/#optional-install-multiple-versions-of-r
 
 At various times in my life I've use each of these approaches for an extended period of time, each works well. Today I use rig.

 
 
## Switching between platforms:

Keras saved models (with the `.keras` extension) are safe to transfer between platforms, and also, between different R versions. Note that the `.keras` extension does not bundle any R code, so if there are custom objects contained in your saved model, you will need to provide them either in the `keras3::load_model(,custom_objects=)` call, use the scope manager `keras3::with_custom_object_scope()`, or register them with in the keras session with `keras3::register_keras_serializable()` before calling `keras3::load_model()`. At some point this might change, as in theory there is nothing preventing us from augmenting the '.keras' bundle with an `.rds` file that contains the custom R objects, but that isn't implemented today.

That said, one common hurdle when switching between platforms is the GPU. Today, the `tensorflow-metal` package is not compatible with Keras 3, meaning there is no way to use the Arm Mac GPU with Tensorflow and Keras 3. This means that if you write any code that targets the existence of a specific device (e.g., `with(tf$device("gpu:0"), ...)`), that will end up being platform specific. 

Another potential hurdle will be the backend. It's likely that in the near future, Keras will gain a new "MLX" backend that will only work on Arm Macs (https://github.com/keras-team/keras/tree/mlx). Also, it's possible that the default backend on Linux will switch at some point from tensorflow to jax. For this reason, the recommendation is to write backend agnostic code as much as possible. Prefer using the `keras3::op_*` family over using the tensorflow backed directly.

**Reply by jonbry** [reply]
Wow, this is absolutely fantastic! Thank you so much for taking the time to put all of this information together! I am just about up and running and this made everything a breeze so far. I hadn't considered issues with different backends, so I'll need to update things to make them a bit more future proof/agnostic.

How did I not know about rig? This makes managing R installations so much easier.
